docs/design/2019-11-05-index-advisor.md

```diff
@@ -0,0 +1,160 @@
+
+- [IBM-DB2 index advisor](https://www.ibm.com/support/knowledgecenter/en/SSEPGG_10.1.0/com.ibm.db2.luw.admin.cmd.doc/doc/r0002452.html)
+
+## Compatibility
+
+Above discussion highlights the fact that as a building block of index advisor, TiDB SQL  optimizer needs to provide following three functionalities for index advisor: 
+
+1. **add hypothetic index**. That is, adding some *virtual indexes* to the schema in memory, which in the view of SQL optimizer, seems to be real indexes thus the optimizer will output execution plan with a vision that these virtual indexes do actually exist.  
+
+2. **alter invisible index**. There are situations where some already existed indexes have little affect on given workload. At this time, we need to consider if these *outdated* indexes should be removed. Instead of deleting these indexes inconsiderately, we can make our decisions after observing the query efficiency with these indexes being muted. To achieve this, we could append a keyword *invisible* to target index's metadata in schema to mute it. 
+
+3. **get plan and its cost**. Best physical plan and its cost output by SQL optimizer are essentials to index advisor to recommend indexes for given workload.  
+
+
+## Implementation
+
+### <span id="compatibility">**Support from TiDB SQL optimizer** </span>
+
+1. **add hypothetic indexes to schema** 
+   - **effect**: query should be compiled with *schema* that contains input hypothetic indexes 
+   - **requirements**: these hypothetic indexes should take effects only within current index advisor session  
+   - **implementation**: get a copy of *session schema* in memory. Then add hypothetic indexes to schema copy and compile the query with this modified schema copy
+
+2. **alter invisible indexes**
+   - **effect**: input existed indexes on table should be invisible to optimizer 
+   - **requirements**: only within current index advisor session, input existed indexes are invisible to optimizer 
+   - **implementation**: get a copy of *session schema* in memory. Then delete input indexes from schema copy and compile the query with this modified schema copy
+
+3. **extract LSIs and cost from execution plan**
+   - **effect**: given the execution plan output by SQL optimizer, related indexes in this plan should be extracted as local selected indexes. Besides, best plan's `cost` should be returned to the top layer of index advisor that invokes TiDB's SQL optimizer.
+   - **requirements**: no
+   - **implementation**: maybe rewrite the function `compiler.Compile` is a simpler way
+
+### **Implement top layer logic of index advisor**
+
+Add a new SQL grammar for index advisor:
+
+```
+INDEX ADVISE
+  [LOCAL]
+  INFILE 'file_name'
+  [MAX_MINUTES number]
+  [MAX_IDXNUM
+     [PER_TABLE number] 
+     [PER_DB number]
```

Comment:
What does this field design for?

Reply:
Maximum allowed recommended indexes. Sometimes DBA doesn't want the number of indexes on specific table exceeds some number, while sometimes the sum of all tables in a database matters. This option provides flexibility for DBAs to consider index recommendation from either the view of each table or the whole database.